crates/sui-bridge/src/action_executor.rs

```diff
@@ -0,0 +1,894 @@
+        execution_queue_sender: mysten_metrics::metered_channel::Sender<
+            VerifiedCertifiedBridgeAction,
+        >,
+    ) {
+        let BridgeActionExecutionWrapper(action, attempt_times) = action;
+        // TODO: use different threshold based on action types.
+        match auth_agg
+            .request_committee_signatures(action.clone(), VALIDITY_THRESHOLD)
+            .await
+        {
+            Ok(certificate) => {
+                execution_queue_sender
+                    .send(certificate)
+                    .await
+                    .expect("Sending to execution queue should not fail");
+            }
+            Err(e) => {
+                warn!("Failed to collect sigs for bridge action: {:?}", e);
+
+                // delay schedule: at most 16 times including the initial attempt
+                // 0.1s, 0.2s, 0.4s, 0.8s, 1.6s, 3.2s, 6.4s, 12.8s, 25.6s, 51.2s, 102.4s, 204.8s, 409.6s, 819.2s, 1638.4s
+                if attempt_times >= 15 {
+                    error!("Manual intervention is required. Failed to collect sigs for bridge action after 16 attempts: {:?}", e);
+                    return;
+                }
+                let delay_ms = 100 * (2 ^ attempt_times);
+                tokio::time::sleep(tokio::time::Duration::from_millis(delay_ms)).await;
+                signing_queue_sender
+                    .send(BridgeActionExecutionWrapper(action, attempt_times + 1))
+                    .await
+                    .expect("Sending to signing queue should not fail");
+            }
+        }
+    }
+
+    // Before calling this function, `key` and `sui_address` need to be
+    // verified to match.
+    async fn run_onchain_execution_loop(
+        sui_client: Arc<SuiClient<C>>,
+        sui_key: SuiKeyPair,
+        sui_address: SuiAddress,
+        mut gas_object_ref: ObjectRef,
+        store: Arc<BridgeOrchestratorTables>,
+        execution_queue_sender: mysten_metrics::metered_channel::Sender<
+            VerifiedCertifiedBridgeAction,
+        >,
+        mut execution_queue_receiver: mysten_metrics::metered_channel::Receiver<
+            VerifiedCertifiedBridgeAction,
+        >,
+    ) {
+        info!("Starting run_onchain_execution_loop");
+        while let Some(certificate) = execution_queue_receiver.recv().await {
+            info!("Received certified action for execution: {:?}", certificate);
+            let tx_data = build_transaction(&gas_object_ref);
+            let sig = Signature::new_secure(
+                &IntentMessage::new(Intent::sui_transaction(), &tx_data),
+                &sui_key,
+            );
+            let signed_tx = Transaction::from_data(tx_data, vec![sig]);
+            let tx_digest = *signed_tx.digest();
+            info!(?tx_digest, ?gas_object_ref, "Sending transaction to Sui");
+            // TODO: add metrics to detect low balances and so on
```

Comment:
In order to do so you would end up querying the latest object state anyway. So I think doing that as part of the retry loop will simplify some of your logic